### PR TITLE
Fixes metric for counting number of append vecs open as file io

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

The metrics for tracking the number of append vecs open as file io is busted.

We currently always decrement the counter on `Drop`, but only _conditionally_ increment the counter if we create a new append vec and write to it first. This means at startup when we open existing append vecs as file io, they do *not* increment the counter, but they *do* decrement it when dropped (like from shrink).

So we end up with metrics like this:
Screenshot 2024-07-10 at 1 13 31 PM
where the counts are negative.

#### Summary of Changes

Always increment the counter when opening an append vec from file when using file io.

### CODE REVIEW

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -1113,7 +1113,6 @@ impl AccountStorageEntry {
             return None;
         }
 
-        APPEND_VEC_REOPEN_AS_FILE_IO.fetch_add(1, Ordering::Relaxed);
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs:1116]: Why is this removed?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:1116]: Good eye! This was moved into `AppendVec::new_from_file_unchecked()`.

The current impl only increments the counter here when calling `reopen_as_readonly()`. By moving the increment to `new_from_file_unchecked()`, we can track all file io append vecs.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:1116]: https://github.com/anza-xyz/agave/blob/53e839c5915c8ba9bc86de3b6a89c389c3f3e3bf/accounts-db/src/accounts_db.rs#L4225
https://github.com/anza-xyz/agave/blob/53e839c5915c8ba9bc86de3b6a89c389c3f3e3bf/accounts-db/src/accounts_db.rs#L5700
https://github.com/anza-xyz/agave/blob/53e839c5915c8ba9bc86de3b6a89c389c3f3e3bf/accounts-db/src/accounts_db.rs#L1088
https://github.com/anza-xyz/agave/blob/53e839c5915c8ba9bc86de3b6a89c389c3f3e3bf/accounts-db/src/accounts_file.rs#L316

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:1116]: It seems for reopen_as_readonly(), the store was created with `AppendVec::new`, not `new_from_file_unchecked()`. 
When this store is created, only *APPEND_VEC_MMAPPED_FILES_OPEN* increased. Therefore, when we reopen it as file_io, we should increase `APPEND_VEC_REOPEN_AS_FILE_IO` here?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:1116]: Yes, correct, we want to increment `APPEND_VEC_REOPEN_AS_FILE_IO` when reopening as file io.

That'll still happen with this PR. `reopen_as_readonly()` calls `new_from_file()` (and then `new_from_file_unchecked()`) with the storage access parameter set to `File`. This increments the counter.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:1116]: I see. You are right. APPEND_VEC_REOPEN_AS_FILE_IO would be correctly increased by 1 with the PR. 
What about  `APPEND_VEC_MMAPPED_FILES_OPEN`?
It looks like we double increase `APPEND_VEC_MMAPPED_FILES_OPEN`. One from AppendVec::new(), and One from new_from_file(). 

Should we decrease ` APPEND_VEC_MMAPPED_FILES_OPEN.fetch_sub(1, Ordering::Relaxed);` here?

Comment (brooksprumo) [reply accounts-db/src/accounts_db.rs:1116]: I think the currently code will correctly decrement the counter.

Since `reopen_as_readonly()` creates a new AccountStorageEntry, when the current/existing/previous one is dropped, the Drop impl on AppendVec will properly decrement the counter. So it's a bit unusual, as we'll have two append vecs open concurrently pointing to the same file, but that should be temporary.

Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs:1116]: yeah. you are right!

**Review by HaoranYi — APPROVED** [review]
lgtm
:shipit:
